shell/packages/sandstorm-db/db.js

```diff
@@ -560,6 +566,52 @@ isSignedUpOrDemo = function () {
   return false;
 }
 
+var calculateReferralBonus = function(accountId, plan) {
+  // This function returns an object of the form:
+  //
+  // - {grains: 0, storage: 0}
+  //
+  // which are extra resources this account gets as part of participating in the referral
+  // program. (Storage is measured in bytes, as usual for plans.)
+
+
+  // Authorization note: Only call this if accountId is the current user!
+  var isPaid = (plan && plan !== "free");
+
+  successfulReferralsCount = countReferrals(accountId);
+  if (isPaid) {
+    var maxPaidStorageBonus = 30 * 1e9;
+    return {grains: 0,
+            storage: Math.min(
+              successfulReferralsCount * 2 * 1e9,
+              maxPaidStorageBonus)};
+  } else {
+    var maxFreeStorageBonus = 2 * 1e9;
+    return {grains: successfulReferralsCount * Infinity,
+            storage: Math.min(
+              successfulReferralsCount * 50 * 1e6,
+              maxFreeStorageBonus)};
+  }
+}
+
+var countReferrals = function (accountId) {
+  var referredIdentityIds = Meteor.users.findOne(
+    {_id: accountId},
+    {fields: {referredIdentityIds: 1}}).referredIdentityIds;
+  return (referredIdentityIds && referredIdentityIds.length || 0);
+}
+
+getUserQuota = function (user) {
+  // Re-fetch the user, since typically Meteor gives is an object with just an _id.
```

Comment:
Fair enough. I guess I'm a bit freaked-out by the fact that this function is available (as far as I can tell?) on both the server & client. I might not know the full server vs. client rules, though. And certainly querying against `Meteor.users` on the client will always get exactly what is published, so re-fetching against that is meaningless on the client.

Fixed in a follow-up commit.